tests/end_to_end/CairoPrecompiles/test_batch_cairo_precompile.py

```diff
@@ -0,0 +1,147 @@
+from typing import Any, List, Tuple
+
+import pytest
+import pytest_asyncio
+from eth_abi import encode
+from eth_utils import keccak
+from hypothesis import given, settings
+from hypothesis import strategies as st
+
+from kakarot_scripts.utils.kakarot import deploy, eth_send_transaction
+from kakarot_scripts.utils.starknet import get_contract, wait_for_transaction
+from tests.utils.errors import cairo_error
+
+EVM_MULTICALLCAIRO_SELECTOR = keccak(
+    text="call_contract(uint256,uint256,uint256,uint256[])"
+)[:4]
+
+
+@pytest_asyncio.fixture(scope="module")
+async def cairo_counter(max_fee, deployer):
+    cairo_counter = get_contract("Counter", provider=deployer)
+
+    yield cairo_counter
+
+    tx = await cairo_counter.functions["set_counter"].invoke_v1(0, max_fee=max_fee)
+    await wait_for_transaction(tx.hash)
+
+
+@pytest_asyncio.fixture(scope="module")
+async def multicall_cairo_counter_caller(owner, cairo_counter):
+    caller_contract = await deploy(
+        "CairoPrecompiles",
+        "MulticallCairoCounterCaller",
+        cairo_counter.address,
+        caller_eoa=owner.starknet_contract,
+    )
+    return caller_contract
+
+
+def encode_starknet_call(call) -> bytes:
+    return encode(
+        ["uint256", "uint256", "uint256[]"],
+        [call.to_addr, call.selector, call.calldata],
+    )
+
+
+def prepare_transaction_data(calls: List[Tuple[Any, str, List[Any]]]) -> str:
+    encoded_calls = b"".join(
+        encode_starknet_call(entrypoint.prepare_call(*calldata))
+        for entrypoint, calldata in calls
+    )
```

Comment:
```suggestion
    encoded_calls = b"".join(
        encode_starknet_call(call)
        for call in calls
    )
```

Reply:
not sure I follow this. why don't you prepare the call?